debian/mindycomp.1
.\" @(#)mindycomp.1		-*- nroff -*-
.TH mindycomp 1 "03 March 2001" "Gwydion Dylan" "Gwydion Dylan"
.UC
.SH NAME
mindycomp \- A Dylan Byte-code Compiler
.SH SYNOPSIS
.B mindycomp
.B [\| -D[Options] \|]
.I input-file\c
.I output-file\c
\|]
.SH WARNING
This man page is intended to supplement the main Gwydion Dylan
documentation at
.IR http://www.gwydiondylan.org/ .
It provides an incomplete reference of the available options. For a full
reference or a tutorial, please see the web site.
.SH DESCRIPTION
.B mindycomp
is a batch compiler for the Mindy Dylan interpreter.  It accepts a subset
of the full Dylan language, and compiles it into a bytecode representation
that can be run using the Mindy(1) interpreter.

.SH SEE ALSO
mindy(1), mindyexec(1), dylan(7), gwydion(7).
.PP
Much more extensive documentation in various formats is available through
the Gwydion Dylan web site at
.IR http://www.gwydiondylan.org/ .
.SH AUTHORS
The Gwydion Group at CMU and the current Gwydion Dylan maintainers. See the
CREDITS file in the Gwydion Dylan distribution for more information.



NEWS
Gwydion Dylan 2.4.0 (2004-10-23)

  Because the bulk of our recent development effort has been put into
  the 2.5 development branch, work on this branch has focused on
  stabilization in the form of bug and compatibility fixes.  Changes
  include improvements in the robustness of the melange tool, and an
  update of the Macintosh CodeWarrior plugin.

Gwydion Dylan 2.3.11 (2003-07-28)

  This release features major work in floating-point functionality,
  macro hygiene, and common-dylan compliance and feature support,
  along with further improvements in introspection support and the
  build system.  We have added some more workarounds for buggy code
  generation in some versions of gcc, and of course we've fixed a few
  of our own bugs too.

Gwydion Dylan 2.3.10 (2003-04-12)

  For this release, we have further improved Common Dylan compliance,
  exception reporting, the Melange tool, introspection, and runtime error
  checking.  We have removed the Boehm/Demers/Weiser garbage collector from
  our distribution, and rely on an externally-provided version.  There
  are workarounds for problems encountered with some versions of gcc 3.x.
  Support for the HP/UX PA-RISC platform has been revived, and an alpha
  version of support for the MacOS X Cocoa framework is now included.

  There's a technology demonstration of a Dylan interactor (also known
  as a REPL to the Lisp folks). Try 'd2c -i'.

Gwydion Dylan 2.3.9 (2002-10-17)

  Besides the obvious ton of bugfixes, we have now working class and
  each-subclass slots, a rudimentary network library, greatly improved
  common-dylan compliance, better performance, increased
  debuggability, updated documentation and all that. 2.3.9 is probably
  the last unstable release before 2.4. All we want to do before 2.4
  is making all the quality assurance tests pass.

Gwydion Dylan 2.3.8 (2002-03-05)

  More bug fixes.

Gwydion Dylan 2.3.7 (2002-02-21)

  This release mainly fixes a bug in class slots, and is considered
  a maintenance release. Various other tweaks and bugfixes as well.

Gwydion Dylan 2.3.6 (2001-11-25)

  The most visible change in 2.3.6 is support for DUIM, a framework
  for user interfaces.  It's supposed to be portable between
  platforms, the version we ship is based on GTK+.  Many thanks to
  Functional Objects, who provided most of their DUIM implementation
  as Open Source to us.  Be warned, our DUIM is still way beta.

  Other changes include ports to MacOS X and Linux on S/390, a ton of
  bugfixes too numerous to mention in detail, support for class slots,
  inlining (non-gf) functions.  Have a look at the ChangeLog for all
  the gory details.

Gwydion Dylan 2.3.5 (2001-05-26)

  Some measurable performance and build-time optimizations have been
  incorporated into d2c.  The compiler now supports limited containers,
  and a bug in class variable initialization has been fixed.  The
  Common Dylan library implementation has expanded further, and support
  for the myriad styles of end-of-line delimiters has been strengthened.

Gwydion Dylan 2.3.4 (2000-11-30)

  The cygnus port created in 2.3.3 has now been integrated into the
  build environment (meaning, there's no special incantations needed
  to compile the cygnus port from a previous revision's port).  The
  file-system library (previously a group of no-ops with some limited
  functionality) is now 75% in compilance with the Common-Dylan
  specification from Functional Objects, inc.  The Time library now
  works on most platforms (including cygnus and solaris, two platforms
  previously not supported).  Melange now optionally creates Dylan-
  style virtual structures that shadow the generated C-style structs.
  This, combined with the new ability for the runtime to provide
  slot names of instances, allows for Dylan-style introspection of
  C-header information.  Various bugfixes, including --no-binaries
  for d2c. Gwydion Dylan now builds fully on MacOS X, and features a
  Carbon interface library with several examples including a full GUI
  application.

Gwydion Dylan 2.3.3 (2000-04-12)

  This release is the first official version that includes MacOS
  support for both mindy and d2c. MacOS X mindy is also
  working. Endless bugs have been fixed, and we're moving towards
  common-dylan compliance (this eventually should give you full source
  code compatibility with Functional Developer).

Gwydion Dylan 2.3.2 (1999-08-25)

  New in this release is shared library support, being made possible
  by Peter Housels C heap generator.

Gwydion Dylan 2.3.1 (1999-06-08)

  This is the first release of the new development branch. Lots of
  things changed: the build system has been completely revamped by
  Eric Kidd to support bootstrapping under various conditions, the
  instance? test has been made faster by Andreas Bogk, there's an
  OpenBSD port by Brad Dominy, work has been done (but not yet
  finished) on a Harlequin Dylan compatible c-ffi by Peter Housel,
  Eric Kidd started work on replacing melange by pidgin. There's now
  support for Common Dylan (which includes a file-system and an
  operating-system library by Tom Emerson), which should give source
  code compatibility with Harlequin; dig, a debugging frontend to gdb
  which understands Dylan now works on Posix systems, Jeff Dubrule has
  provided a mindy port to Linux on Alpha, Tom Emerson implemented
  Unicode escape in strings, and of course bugs have been fixed by
  various people.

Gwydion Dylan 2.2.0 (1999-02-10)

  This is basically a polished-up 2.1.3, with several more bugfixes.
  It is intended to be a stable release without too much new code
  introduced.

Gwydion Dylan 2.1.3 (1998-12-??)

  This release includes Melange enhancements by Peter Housel, Andreas Bogk
  and Eric Kidd; bug fixes by Andreas Bogk; optimizer fixes by Matthias
  Hoelzl; and a new parse-arguments library (and lots of documentation) by
  Eric Kidd.

Gwydion Dylan 2.1.2 (1998-12-01)


mindy/comp/mindycomp.c
/**********************************************************************\
*
*  Copyright (c) 1994  Carnegie Mellon University
*  All rights reserved.
*  
*  Use and copying of this software and preparation of derivative
*  works based on this software are permitted, including commercial
*  use, provided that the following conditions are observed:
*  
*  1. This copyright notice must be retained in full on any copies
*     and on appropriate parts of any derivative works.
*  2. Documentation (paper or online) accompanying any system that
*     incorporates this software, or any part of it, must acknowledge
*     the contribution of the Gwydion Project at Carnegie Mellon
*     University.
*  
*  This software is made available "as is".  Neither the authors nor
*  Carnegie Mellon University make any warranty about the software,
*  its performance, or its conformity to any specification.
*  
*  Bug reports, questions, comments, and suggestions should be sent by
*  E-mail to the Internet address "gwydion-bugs@cs.cmu.edu".
*
***********************************************************************
*
* $Header: /scm/cvs/src/mindy/comp/mindycomp.c,v 1.3 1998/12/17 11:03:12 igor Exp $
*
* This file is the main driver.
*
\**********************************************************************/

#include "../compat/std-c.h"
#include "../compat/std-os.h"

#include "mindycomp.h"
#include "parser.h"
#include "src.h"
#include "print.h"
#include "expand.h"
#include "envanal.h"
#include "lexer.h"
#include "header.h"
#include "sym.h"
#include "info.h"
#include "compile.h"
#include "dump.h"
#include "feature.h"
#include "lose.h"

struct body *Program = NULL;

struct symbol *LibraryName = NULL;
struct symbol *ModuleName = NULL;
boolean GiveWarnings = TRUE;

char *current_file = "<stdin>";

static int nerrors = 0;

static void verror(int line, char *msg, va_list ap)
{
    fprintf(stderr, "%s:%d: error: ", current_file, line);
    vfprintf(stderr, msg, ap);
    if (msg[strlen(msg)-1] != '\n')
	putc('\n', stderr);

    nerrors++;
}
#if _USING_PROTOTYPES_
void error(int line, char *msg, ...)
{
    va_list ap;

    va_start(ap, msg);
    verror(line, msg, ap);
    va_end(ap);
}
#else
void error(va_alist) va_dcl
{
    va_list ap;
    int line;
    char *msg;

    va_start(ap);
    line = va_arg(ap, int);
    msg = va_arg(ap, char *);
    verror(line, msg, ap);
    va_end(ap);
}
#endif

static void vwarn(int line, char *msg, va_list ap)
{
    if ( ! GiveWarnings)
        return;
    fprintf(stderr, "%s:%d: warning: ", current_file, line);
    vfprintf(stderr, msg, ap);
    if (msg[strlen(msg)-1] != '\n')
	putc('\n', stderr);
}

#if _USING_PROTOTYPES_
void warn(int line, char *msg, ...)
{
    va_list ap;

    va_start(ap, msg);
    vwarn(line, msg, ap);
    va_end(ap);
}
#else
void warn(va_alist) va_dcl
{
    va_list ap;
    int line;
    char *msg;

    va_start(ap);
    line = va_arg(ap, int);
    msg = va_arg(ap, char *);
    vwarn(line, msg, ap);
    va_end(ap);
}
#endif

static void usage(void)
{
    fprintf(stderr, "usage: mindycomp [-d[p][e]] [-l library-name] "
	    "[-o object-name] [-Dfeature] [-Ufeature] source-name\n");
    exit(1);
}

static void set_module(char *value)
{
    if (ModuleName) {
	fprintf(stderr, "multiple module: file headers.\n");
	exit(1);
    }
    ModuleName = symbol(value);
}

static void set_library(char *value)
{
    if (LibraryName != NULL && strcasecmp((char *)LibraryName->name, value)) {
	fprintf(stderr,
		"Library name specified on the command line differs from\n"
		"the library name specified in the file.\n");
	exit(1);
    }
    if (LibraryName == NULL)
	LibraryName = symbol(value);
}

static void end_of_headers(char *value)
{
    if ( ! ModuleName) {
        warn(line_count-1, "no module: header, assuming Dylan-User\n");
	ModuleName = sym_DylanUser;
    }
}

static char *find_extension(char *source)
{
    char *slash = strrchr(source, '/');
    char *dot = strchr(slash ? slash : source, '.');

    if (dot)
	return dot+1;
    else
	return NULL;
}

static char *make_output_name(char *source, char *new_extension)
{
    char *extension = find_extension(source);
    int base_len = extension ? extension - source - 1 : strlen(source);
    char *output = malloc(base_len + strlen(new_extension) + 1);

    memcpy(output, source, base_len);
    strcpy(output + base_len, new_extension);

    return output;
}

int main(int argc, char *argv[])
{
    boolean print_parse = FALSE;
    boolean print_expanded = FALSE;
    char *arg;
    char *source_name = NULL;
    char *output_name = NULL;
    FILE *file;

    add_header_handler("module", set_module);
    add_header_handler("library", set_library);
    add_header_handler(NULL, end_of_headers);

    init_sym_table();
    init_feature();
    init_info();
    init_expand();
    init_compile();

    while ((arg = *++argv) != NULL) {
	if (arg[0] == '-') {
	    switch (arg[1]) {
	      case 'd':
		if (arg[2] == '\0') {
		    print_parse = TRUE;
		    print_expanded = TRUE;
		}
		else {
		    char *ptr;
		    for (ptr = arg+2; *ptr != '\0'; ptr++) {
			switch (*ptr) {
			  case 'p':
			    print_parse = TRUE;
			    break;
			  case 'e':
			    print_expanded = TRUE;
			    break;
			  default:
			    fprintf(stderr, "Invalid thing to print: ``%c''\n",
				    *ptr);
			    usage();
			}
		    }
		}
		break;

	      case 'o':
		if (output_name != NULL) {
		    fprintf(stderr, "-o can only be used once.\n");
		    usage();
		}
		else if (arg[2] != '\0')
		    output_name = arg+2;
		else if (*++argv == NULL) {
		    fprintf(stderr, "-o must be followed by the output "
			    "file name.\n");
		    usage();
		}
		else
		    output_name = *argv;
		break;

	      case 'l':
		if (LibraryName != NULL) {
		    fprintf(stderr, "-l can only be used once.\n");
		    usage();
		}
		if (arg[2] != '\0')
		    LibraryName = symbol(arg+2);
		else if (*++argv == NULL) {
		    fprintf(stderr, "-l must be followed by the library "
			    "name.\n");
		    usage();
		}
		else
		    LibraryName = symbol(*argv);
		break;

	      case 'q':
		GiveWarnings = FALSE;
		break;

	      case 'D':
		if (arg[2] != '\0')
		    add_feature(symbol(arg + 2));
		else if (*++argv == NULL) {
		    fprintf(stderr,
			    "-D must be followed by the feature to define.\n");
		    usage();
		}
		else
		    add_feature(symbol(*argv));
		break;

	      case 'U':
		if (arg[2] != '\0')
		    remove_feature(symbol(arg + 2));
		else if (*++argv == NULL) {
		    fprintf(stderr, "-U must be followed by the feature "
			    "to undefine.\n");
		    usage();
		}
		else
		    remove_feature(symbol(*argv));
		break;

	      default:
		fprintf(stderr, "Invalid flag: ``%s''\n", arg);
		usage();
	    }
	}
	else if (source_name != NULL) {
	    fprintf(stderr, "Too many files\n");
	    usage();
	}
	else
	    source_name = arg;
    }

    if (source_name == NULL)
	usage();


    yyin = fopen(source_name, "r");
    if (yyin == NULL) {
        /* Try the same filename but in the current directory */
        /* Start ptr at the null termination, and work backwards to a 
	   path separator. */
        char *ptr = source_name + strlen(source_name);
        for ( ; ptr != source_name; ptr--) {
	    if (*ptr == '/' || *ptr == '\\') {
	        /* We're pointing at the path separator, which is too far */
	        ptr++;
	        break;
	    }
	}
	/* If ptr is a different string than source_name, and it isn't
	   the empty string... */
	if (ptr != source_name && *ptr != 0) {
	    yyin = fopen(ptr, "r");
	}
	if (yyin == NULL) {
	    perror(source_name);
	    exit(1);
	}
    }

    current_file = source_name;

    yyparse();

    if (print_parse) {
	printf("================ Original Parse Tree ================\n");
	print_body(Program, 0);
    }

    if (nerrors != 0)
	exit(1);

    /* Do the various source-to-source expansions. */
    expand(Program);

    if (print_expanded) {
	printf("================ Expanded Parse Tree ================\n");
	print_body(Program, 0);
    }

    if (nerrors != 0)
	exit(1);

    /* Run environment analysis */
    environment_analysis(Program);

    if (nerrors != 0)
	exit(1);

    if (output_name == NULL)
	output_name = make_output_name(source_name, ".dbc");

    if (strcmp(output_name, "-") == 0)
        file = stdout;
    else
        file = fopen(output_name, "wb");

    if (file == NULL) {
	perror(output_name);
	exit(1);
    }

    dump_setup_output(source_name, file);

    /* Generate code. */
    compile(Program);

    dump_finalize_output();

    fclose(file);

    exit(0);
    return 0;
}

/* Try to keep this consistent with interp/mindy.c */

#undef malloc
void *check_malloc(size_t sz, char *file, int line)
{
    void *ret = malloc(sz);
    if (ret == 0) 
	lose("malloc failed -- out of memory in %s line %d", file, line);
    return ret;
}

#undef calloc
void *check_calloc(size_t nobj, size_t sz, char *file, int line)
{
    void *ret = calloc(nobj, sz);
    if (ret == 0) 
	lose("calloc failed -- out of memory in %s line %d", file, line);
    return ret;
}

#undef realloc
void *check_realloc(void *ptr, size_t sz, char *file, int line)
{
    void *ret = realloc(ptr, sz);
    if (ret == 0) 
	lose("realloc failed -- out of memory in %s line %d", file, line);
    return ret;
}

